proactive/apache_httpd.go

```diff
@@ -0,0 +1,339 @@
+		// step 2.8
+		smallestContentLength,
+	}
+	return httpd{filters: filters}
+}
+
+func (c httpd) Choose(
+	r *http.Request, reps ...representation.Representation) (representation.Representation, error) {
+
+	var (
+		a   header.Accept
+		ae  header.AcceptEncoding
+		al  header.AcceptLanguage
+		ac  header.AcceptCharset
+		err error
+	)
+
+	accept := r.Header["Accept"]
+	if a, err = header.NewAccept(accept); err != nil {
+		return nil, err
+	}
+
+	acceptEncoding := r.Header["Accept-Encoding"]
+	if ae, err = header.NewAcceptEncoding(acceptEncoding); err != nil {
+		return nil, err
+	}
+
+	acceptLanguage := r.Header["Accept-Language"]
+	if al, err = header.NewAcceptLanguage(acceptLanguage); err != nil {
+		return nil, err
+	}
+
+	acceptCharset := r.Header["Accept-Charset"]
+	if ac, err = header.NewAcceptCharset(acceptCharset); err != nil {
+		return nil, err
+	}
+
+	var variants internal.VariantSet
+	for _, rp := range reps {
+		qt := c.acceptQuality(rp, a)
+		qc := c.acceptCharsetQuality(rp, ac)
+		ql, los := c.acceptLanguageQuality(rp, al)
+		qe := c.acceptEncodingQuality(rp, ae)
+
+		shouldEliminate :=
+			qt == header.QualityValueMinimum || qc == header.QualityValueMinimum ||
+				qe == header.QualityValueMinimum || ql == header.QualityValueMinimum
+		if shouldEliminate {
+			continue
+		}
+
+		variants = append(variants, internal.Variant{
+			Rep:                rp,
+			SourceQ:            header.QualityValue(rp.SourceQuality()),
+			MediaTypeQ:         qt,
+			CharsetQ:           qc,
+			EncodingQ:          qe,
+			LanguageQ:          ql,
+			LanguageOrderScore: los,
+		})
+	}
+
+	// if there are no eligble variants, we are done.
+	if variants.Empty() {
+		return nil, nil
+	}
+
+	for _, f := range c.filters {
+		if variants, err = f(variants); err != nil {
+			return nil, err
+		}
+		if variants.Size() == 1 {
+			break
+		}
+	}
+	return variants.First().Rep, nil
+}
+
+var (
+	smallestContentLength filter = func(variants internal.VariantSet) (internal.VariantSet, error) {
+		variants.Sort(func(i, j int) bool {
+			// sort smallest to largest.
+			var f, s int
+			if bytes, err := variants[i].Rep.Bytes(); err != nil {
+				f = len(bytes)
+			}
+			if bytes, err := variants[j].Rep.Bytes(); err != nil {
+				s = len(bytes)
+			}
+			return f > s
+		})
+		lowest := variants.First()
+		lowestBytes, err := lowest.Rep.Bytes()
+		if err != nil {
+			return internal.VariantSet{}, err
+		}
+		lowestLength := len(lowestBytes)
+		return variants.Where(func(v internal.Variant) bool {
+			var length int
+			if bytes, err := v.Rep.Bytes(); err == nil {
+				length = len(bytes)
+			}
+			return length == lowestLength
+		}), nil
+	}
+
+	bestEncoding filter = func(variants internal.VariantSet) (internal.VariantSet, error) {
+		variants.Sort(func(i, j int) bool {
+			return variants[i].EncodingQ < variants[j].EncodingQ
+		})
+		highest := variants.First()
+		return variants.Where(func(v internal.Variant) bool {
+			return v.EncodingQ == highest.EncodingQ
```

Comment:
use `Equals(…)`, `LessThan(…)`, or `GreaterThan(…)` instead of `==`, `<`, or `>`.

Reply:
done!